### DISCUSSION TITLE
Any way to get Git Blame to use file modified time for uncommitted lines?

### DISCUSSION THREAD

**Discussion by eamodio** [git-for-windows/git#3065] [Q&A]
Is there any way to get Git blame to use the modified time of the file for uncommitted lines, rather than the time the blame command is run?

**Comment by dscho** [comment]
Internally, `git blame` constructs a virtual commit for those uncommitted files: https://github.com/git/git/blob/v2.30.1/blame.c#L172-L175. This commit will have only one author time, the one that is currently used is `now`.

Since `git blame` [only works on one file at a time](https://github.com/git/git/blob/v2.30.1/builtin/blame.c#L1022-L1030), it seems totally doable to `lstat()` and assign the `st_mtime` to the date to be used. This would have to be done here: https://github.com/git/git/blob/v2.30.1/blame.c#L194.

You could implement this [and contribute it](https://git-scm.com/docs/SubmittingPatches) (most likely as an opt-in feature, to make it easier to accept the change).

In the meantime, if you want to work around the issue, you could fake that commit yourself. Something like this should get you started:

```bash
file=the/file/to/blame
tmp_index=$(git rev-parse --git-path tmp-index.$$)
cp "$(git rev-parse --git-path index)" "$tmp_index"
GIT_INDEX_FILE=$tmp_index git add "$file"
tree=$(GIT_INDEX_FILE=$tmp_index git write-tree)
mtime = $(stat -c %Y "$file") # This would be `stat -f %m "$file" on macOS
commit=$(echo "fake message" | GIT_AUTHOR_DATE=$mtime GIT_COMMITTER_DATE=$mtime git commit-tree $tree -p HEAD)
git blame $file $commit
```

(Note: this is _completely_ untested code, meant only for inspiration, not for mindless copy/pasting)

**Reply by eamodio** [reply]
Unfortunately, my C is extremely rusty, nor do I even that the toolchain to get started on this. Is there a devcontainer or anything where a contributor could easily spin up a pre-configured container for this? Or even a Codespace?

**Reply by dscho** [reply]
I _think_ it should work in a Codespace. Just make sure to run `contrib/vscode/init.sh` to initialize the VS Code configuration.
